6024: DAPI Adresse validering på fatt vedtak

thanglyphan opened the pull request:
https://jira.adeo.no/browse/MELOSYS-6024

thanglyphan committed:
6024: Adresse validering på fatt vedtak

https://jira.adeo.no/browse/MELOSYS-6024

janitamarielle commented on src/ducks/kontroll/operations.ts:
```diff
@@ -7,6 +7,10 @@ export function resetKontroll() {
   return Actions.resetKontroll();
 }
 
+export function kontrollFeil(feilData: any) {
```
Fra navnet skjønner jeg ikke hva denne funksjonen gjør. kontrollerFeil? Hva er `feil` her?
Gjelder også actions.ts.

frans-a replied:
Diskuteres her:
https://github.com/navikt/melosys-web/pull/1978#discussion_r1293050252

janitamarielle commented on src/felleskomponenter/enkelStegvelger/enkelStegvelger.tsx:
```diff
@@ -38,6 +42,8 @@ export default ({ alleSteg }: EnkelStegvelgerProps) => {
       aktuelleSteg?.map((steg: AktueltSteg) => ({ ...steg, aktivtSteg: steg.stegPosisjon === aktivtStegIndex }))
     );
+    dispatch(kontrollOperations.resetKontroll());
+    dispatch(feiletResponsOperations.resetFeiletRespons());
```
Denne skal fjernes hver gang man bytter steg? Trengs det?

thanglyphan replied:
Jeg vet ikke helt om det trengs, men har vært snakket om at feilmeldinger "henger igjen" ofte når man bytter steg :/ Jeg la det til der basert på det egentlig. Hvis det ikke trengs kan det sikkert flyttes til når vedtakssteget unmounter

janitamarielle replied:
Det har tidligere vært saksbehandling-filen sitt ansvar å rydde opp etter en flyt - høres ut som om det kanskje burde flyttes dit?

frans-a replied:
Trenger tips eller howto hvis jeg skal gjøre dette.

janitamarielle replied:
Tenker fjerne disse to linjene, og så dobbeltsjekke at de allerede ligger i onUnMount/useEffect i de saksbehandling-filene som bruker enkelStegvelger (unntaksregistrering, ikke-yrkesaktiv, og FTRL)

janitamarielle commented on src/felleskomponenter/stegvelger/Stegvelger.jsx:
```diff
@@ -522,7 +522,8 @@ class Stegvelger extends Component {
         <StegLinje steg={this.state.aktuelleSteg} stegKlikk={this.validerSoknadOgGaTilSteg} />
         {!this.props.redigerbart && <Innsynsmelding />}
-        {this.erVedtakSteg(this.state.aktivtStegNummer) && <Feilmeldinger />}
+        {(this.erVedtakSteg(this.state.aktivtStegNummer) ||
```
Trekke ut i en const visFeilmeldinger?

janitamarielle commented on src/sider/eu_eøs/stegKomponenter/vurderingArtikkel16Anmodning.jsx:
```diff
@@ -279,6 +291,47 @@ class VurderingArtikkel16Anmodning extends Component {
+    const { kontrollFeil, resetKontroll, behandlingID } = this.props;
+    this.setState({ sjekkerAdresse: true });
+    Api.Kontroll.harRegistrertAdresse({
+      brukerID: "",
```
Trenger du sende inn brukerID og orgnr her?

frans-a replied:
Trengs ikke i Javascript men trengs vel i Typescript (strict).  Hva er ønskelig? Skal jeg sette `null`?

janitamarielle replied:
Tenker endre typescript siden det ikke kreves. 
Sette `brukerID?: string | null` og tilsvarenende på de andre parametrene.

janitamarielle commented on src/sider/eu_eøs/stegKomponenter/vurderingArtikkel16Anmodning.jsx:
```diff
@@ -279,6 +291,47 @@ class VurderingArtikkel16Anmodning extends Component {
+            {
+              felter: [],
+              kode:
+                res.rolle === MKV.Koder.aktoersroller.BRUKER
```
Er ikke et must å endre men vurder gjerne: Hvis rolle sendes kun for å bestemme teksten på kontrollfeilen, er det verdt å se på å sende liste med feil-koder(kontroll_begrunnelser) tilsvarende de andre kontroll-endepunkten? Så er tom liste == ingen feil, og liste med element så bruker du elementet til å sette kontroll-feilen?

frans-a replied:
Helt enig. Det er ikke lett forståelig å sende rolle. Skal sende feilkode i stedet

janitamarielle commented on src/sider/eu_eøs/stegKomponenter/vurderingArtikkel16Anmodning.jsx:
```diff
@@ -279,6 +291,47 @@ class VurderingArtikkel16Anmodning extends Component {
+            kontrollfeilList,
+          });
+        } else {
+          resetKontroll();
```
resetKontrollFeil?

frans-a replied:
Det blir nok avhengig av hvordan vi kaller `kontrollOperations.kontrollFeil`

janitamarielle commented on src/sider/eu_eøs/stegKomponenter/vurderingArtikkel16Anmodning.jsx:
```diff
@@ -629,4 +686,9 @@ const VurderingArtikkel16AnmodningForm = reduxForm({
 
-export default connect(mapStateToProps)(VurderingArtikkel16AnmodningForm);
+const mapDispatchToProps = (dispatch) => ({
+  kontrollFeil: (kontrollBegrunnelse) => dispatch(kontrollOperations.kontrollFeil(kontrollBegrunnelse)),
```
er dette egentlig oppdaterKontrollFeil?

frans-a replied:
`kontrollOperations.kontrollFeil` lager bare et objekt som sendes til redux. 
Hvilken navnekonvensjon har vi for dette? Enig i at kontrollFeil er utydelig

janitamarielle replied:
Vi har nok brukt oppdater der det gjelder lokal state (aka redux)

frans-a replied:
Da er `oppdaterKontrollFeil` greit for meg og. Det kunne være kontrollVarsel eller kontrollMelding og, f.eks.

I
```
export function kontrollFeil(feilData: any) {
  return { type: Types.OK, data: feilData };
}
```
er det normalt å sende OK, selv om det er for en feil? Virker ikke intuitivt

janitamarielle replied:
KontrollFeil er allerede et begrep som brukes flere steder, så det tenker jeg er greit nok. 

Hmm, nei her har man tilsynelatende gjenbrukt den action-typen som brukes av "respons fra backend ved kontroll gikk bra"-actionen. Vi vil nok lage en ny action: Types.OPPDATER_KONTROLLFEIL og legge inn støtte for dette i kontroll/reducers.js ++.

janitamarielle commented on src/sider/eu_eøs/stegKomponenter/vurderingArtikkel16Anmodning.jsx:
```diff
@@ -279,6 +291,47 @@ class VurderingArtikkel16Anmodning extends Component {
+        const kontrollfeilList = [
+          {
+            felter: [],
+            kode: MKV.Koder.begrunnelser.kontroll_begrunnelser.MANGLENDE_REGISTRERTE_ADRESSE_BRUKER,
```
Hvorfor denne koden i kontrollfeilList? Kan man ikke bruke erroren som kommer fra catch eller noe mer generisk "Noe teknisk gikk feil, try again"? Mistenker dette kan bli misvisende.

frans-a replied:
Har du et forslag på hvordan vi kan løse dette? Enig at det kan være misvisende, som vi ser ved journalføring

frans-a committed:
Mer konsekvent navngivning

frans-a committed:
Lesbarhet

frans-a committed:
Argumenter ikke påkrevd

frans-a committed:
Konsekvent navn

frans-a committed:
Refaktorering api

journalføring

janitamarielle commented on src/sider/eu_eøs/saksbehandling/saksbehandling.jsx:
```diff
@@ -332,7 +332,7 @@ const mapDispatchToProps = (dispatch) => ({
     dispatch(anmodningsperiodesvarOperations.hent(anmodningsperiodeID)),
   resetFeiletrespons: () => dispatch(feiletResponsOperations.resetFeiletRespons()),
-  resetKontroll: () => dispatch(kontrollOperations.resetKontroll()),
+  resetKontroll: () => dispatch(kontrollOperations.resetKontrollFeil()),
```
rename funksjonen vi lager her også?

frans-a committed:
Refaktorering api

anmodning om unntak

frans-a committed:
Bump kodeverk

barstadEkstern commented on src/sider/eu_eøs/stegKomponenter/vurderingArtikkel16Anmodning.jsx:
```diff
@@ -279,6 +291,36 @@ class VurderingArtikkel16Anmodning extends Component {
+        const kontrollfeilList = [
+          {
+            kode: MKV.Koder.begrunnelser.kontroll_begrunnelser.ANNET, //FIXME hva fungerer her?
+            felter: ["Hjelp, har du et forslag?"],
```
Vi må endre til noe hensiktsmessig her.

janitamarielle replied:
"Vi opplevde en teknisk feil når vi sjekket adressen til brukeren. Dersom problemet vedvarer lag gjerne en vaktsak så skal vi se på problemet."

Men, kanskje dette er en utmerket ting å spørre fag om? :P

frans-a replied:
Tror jeg fant noe hensiktsmessig.
Ville spørre fag hvis det var noe generisk, men her synes jeg ikke at det er verdt det

barstadEkstern reviewed (commented):
Litt diverse som må fikses og en del linting som må ordnes. Kan hjelpe med det på Mandag

frans-a committed:
Linting

janitamarielle commented on src/ducks/kontroll/reducers.ts:
```diff
@@ -21,6 +21,8 @@ export default function reducer(state = initialState, action: Types.Action) {
       return { ...state, status: STATUS.ERROR, data: action.data };
     case Types.OK:
       return { ...state, status: STATUS.OK, data: action.data };
+    case Types.OPPDATER_KONTROLLFEIL:
```
Siden denne kun endrer data, tenker jeg man kan fjerne oppdatering av status her. Og da bare ha `{ ...state, data: action.data } `så forblir status det den var.

janitamarielle commented on src/sider/eu_eøs/stegKomponenter/vurderingArtikkel16Anmodning.jsx:
```diff
@@ -550,7 +592,7 @@ class VurderingArtikkel16Anmodning extends Component {
                 spinner: anmodningPending,
                 autoDisableVedSpinner: true,
-                disabled: !redigerbart,
+                disabled: !redigerbart || this.state.harFeil,
```
Kan hende du vil trekke denne ut sammen med de andre state-verdiene (lengre opp) for å være konsekvent: 
```
const {
      begrunnelserFeilmelding,
      begrunnelseFritekstBrevFeilmelding,
      begrunnelseFritekstSedFeilmelding,
      lovvalgFeilmelding,
      sendBrevFeilmelding,
      anmodningPending,
      valgteVedlegg,
    } = this.state;
```

janitamarielle commented on src/sider/eu_eøs/stegKomponenter/vurderingArtikkel16Anmodning.jsx:
```diff
@@ -279,6 +291,36 @@ class VurderingArtikkel16Anmodning extends Component {
+
+        const kontrollfeilList = [
+          {
+            kode: MKV.Koder.begrunnelser.kontroll_begrunnelser.ANNET, // FIXME hva fungerer her?
```
Må vi sende inn kode i det hele tatt?

frans-a replied:
Nei :)

janitamarielle commented on src/sider/eu_eøs/stegKomponenter/vurderingArtikkel16Anmodning.jsx:
```diff
@@ -279,6 +291,36 @@ class VurderingArtikkel16Anmodning extends Component {
+            felter: ["Hjelp, har du et forslag?"],
+          },
+        ];
+        oppdaterKontrollFeil({ kontrollfeilList });
```
Trekke kontrollfeilList opprettelsen rett inn i denne funksjonen? Syns ikke helt det gir så mye å ha det i en egen const. 🤷‍♀️

janitamarielle commented on src/sider/ftrl/saksbehandling/saksbehandling.tsx:
```diff
@@ -84,7 +84,7 @@ const mapDispatchToProps = (dispatch: ThunkDispatch<RootState, unknown, Action>)
   skjulMenypanel: () => dispatch(menypanelOperations.skjulMenypanel()),
   resetFeiletrespons: () => dispatch(feiletResponsOperations.resetFeiletRespons()),
-  resetKontroll: () => dispatch(kontrollOperations.resetKontroll()),
+  resetKontroll: () => dispatch(kontrollOperations.resetKontrollFeil()),
```
resetKontrollFeil på funksjonen vi oppretter også her

frans-a committed:
QA

frans-a committed:
Feilmelding ved teknisk feil

frans-a committed:
Linting

frans-a committed:
QA

frans-a committed:
Endre ikke status

Siden denne kun endrer data, tenker jeg man kan fjerne oppdatering av status her. Og da bare ha { ...state, data: action.data } så forblir status det den var.

frans-a committed:
For å være konsekvent

frans-a committed:
QA

frans-a commented:
Tatt alle kommentarer og testet en del lokalt. Må finne et godt tidspunkt for å merge siden det blir DAPI + DWEB

frans-a committed:
em greit for border

Ifølge https://github.com/navikt/melosys-devops/blob/main/docs/utvikling-melosys-web.md

janitamarielle commented on src/sider/eu_eøs/stegKomponenter/vurderingArtikkel16Anmodning.jsx:
```diff
@@ -279,6 +291,30 @@ class VurderingArtikkel16Anmodning extends Component {
+      .catch(() => {
+        const feilmelding =
+          "En teknisk feil skjedde da adresser skulle sjekkes. Prøv igjen eller kontakt brukerstøtte hvis problemet vedvarer.";
+        this.setState({ harFeil: true, sendBrevFeilmelding: feilmelding, sjekkerAdresse: false });
```
Skal ikke sendBrevFeilmelding vises et sted?

frans-a replied:
Det vises med
``` js
          {sendBrevFeilmelding && (
            <Nav.AlertStripe type="advarsel" className="varsel">
              {sendBrevFeilmelding}
            </Nav.AlertStripe>
          )}
```
over knappen

janitamarielle replied:
Ah, du brukte eksisterende state 🧠 

Ville kanskje valgt rød/error-melding siden det blokkerer fortsettelse, men er forsåvidt greit.

frans-a replied:
Enig, fikset

janitamarielle reviewed (approved):
Har ikke kjørt opp lokalt (lemme know om du vil jeg skal gjøre det), men det ser ok ut dette 👍

frans-a committed:
rød/error-melding siden det blokkerer fortsettelse

frans-a committed:
Merge branch 'master' into MELOSYS-6024__adr_fix
